CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+
+    # afterwards we have attribute macros
+    attr_reader :name
+
+    # followed by other macros (if any)
+    validates :name
+
+    # public class methods are next in line
+    def self.some_method
+    end
+
+    # initialization goes between class methods and other instance methods
+    def initialize
+    end
+
+    # followed by other public instance methods
+    def some_method
+    end
+
+    # protected and private methods are grouped near the end
+    protected
+
+    def some_protected_method
+    end
+
+    private
+
+    def some_private_method
+    end
+  end
+  ```
+
+* <a name="file-classes"></a>
+  Don't nest multi line classes within classes. Try to have such nested
+  classes each in their own file in a folder named like the containing class.
+<sup>[[link](#file-classes)]</sup>
+
+  ```Ruby
+  # bad
+
+  # foo.rb
+  class Foo
+    class Bar
+      # 30 methods inside
+    end
+
+    class Car
+      # 20 methods inside
+    end
+
+    # 30 methods inside
+  end
+
+  # good
+
+  # foo.rb
+  class Foo
+    # 30 methods inside
+  end
+
+  # foo/bar.rb
+  class Foo
+    class Bar
+      # 30 methods inside
+    end
+  end
+
+  # foo/car.rb
+  class Foo
+    class Car
+      # 20 methods inside
+    end
+  end
+  ```
+
+* <a name="modules-vs-classes"></a>
+  Prefer modules to classes with only class methods. Classes should be used
+  only when it makes sense to create instances out of them.
+<sup>[[link](#modules-vs-classes)]</sup>
+
+  ```Ruby
+  # bad
+  class SomeClass
+    def self.some_method
+      # body omitted
+    end
+
+    def self.some_other_method
+    end
+  end
+
+  # good
+  module SomeModule
+    module_function
+
+    def some_method
+      # body omitted
+    end
+
+    def some_other_method
+    end
+  end
+  ```
+
+* <a name="module-function"></a>
+  Favor the use of `module_function` over `extend self` when you want to turn
+  a module's instance methods into class methods.
+<sup>[[link](#module-function)]</sup>
+
+  ```Ruby
+  # bad
+  module Utilities
+    extend self
+
+    def parse_something(string)
+      # do stuff here
+    end
+
+    def other_utility_method(number, string)
+      # do some more stuff
+    end
+  end
+
+  # good
+  module Utilities
+    module_function
+
+    def parse_something(string)
+      # do stuff here
+    end
+
+    def other_utility_method(number, string)
+      # do some more stuff
+    end
+  end
+  ```
+
+* <a name="liskov"></a>
+  When designing class hierarchies make sure that they conform to the [Liskov
+  Substitution
+  Principle](http://en.wikipedia.org/wiki/Liskov_substitution_principle).
+<sup>[[link](#liskov)]</sup>
+
+* <a name="solid-design"></a>
+  Try to make your classes as
+  [SOLID](http://en.wikipedia.org/wiki/SOLID_\(object-oriented_design\)) as
+  possible.
+<sup>[[link](#solid-design)]</sup>
+
+* <a name="define-to-s"></a>
+  Always supply a proper `to_s` method for classes that represent domain
+  objects.
+<sup>[[link](#define-to-s)]</sup>
```

Comment:
The likelihood of everyone actually doing this is low because the utility is low. It'd be cool if we all did this, but it isn't free and the benefits are negligible.